src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,323 @@ pub mod policy;
 pub mod trust_store;
 pub mod types;
 
+use std::collections::HashSet;
+
+use crate::certificate::{cert_is_self_issued, cert_is_self_signed};
+use crate::types::{DNSConstraint, IPAddress, IPConstraint};
+use crate::ApplyNameConstraintStatus::{Applied, Skipped};
+use cryptography_x509::extensions::{DuplicateExtensionsError, Extensions};
+use cryptography_x509::{
+    certificate::Certificate,
+    extensions::{NameConstraints, SubjectAlternativeName},
+    name::GeneralName,
+    oid::{NAME_CONSTRAINTS_OID, SUBJECT_ALTERNATIVE_NAME_OID},
+};
+use ops::CryptoOps;
+use policy::Policy;
+use trust_store::Store;
+use types::DNSName;
+
+#[derive(Debug, PartialEq, Eq)]
 pub enum ValidationError {
-    Other(&'static str),
+    CandidatesExhausted(Box<ValidationError>),
+    Malformed(asn1::ParseError),
+    DuplicateExtension(DuplicateExtensionsError),
+    Other(String),
+}
+
+impl From<asn1::ParseError> for ValidationError {
+    fn from(value: asn1::ParseError) -> Self {
+        Self::Malformed(value)
+    }
+}
+
+impl From<DuplicateExtensionsError> for ValidationError {
+    fn from(value: DuplicateExtensionsError) -> Self {
+        Self::DuplicateExtension(value)
+    }
+}
+
+pub struct Intermediates<'a>(HashSet<Certificate<'a>>);
+
+impl<'a> Intermediates<'a> {
+    fn new<B: CryptoOps>(
+        intermediates: impl IntoIterator<Item = Certificate<'a>>,
+        policy: &Policy<'_, B>,
+    ) -> Result<Self, ValidationError> {
+        Ok(Self(
+            intermediates
+                .into_iter()
+                .map(
+                    |intermediate| match cert_is_self_signed(&intermediate, &policy.ops) {
+                        true => Err(ValidationError::Other(
+                            "self-signed certificate cannot be an intermediate".to_string(),
+                        )),
+                        false => Ok(intermediate),
+                    },
+                )
```

Comment:
Yeah, it causes a potential cycle: if the user passes a self-signed cert in the intermediate set, we'll end up recursing infinitely (since self-signed certs don't increase the validation depth per 5280).

Another option here would be to instead have a separate "physical" depth counter that doesn't follow the 5280 counting rules, and instead gives up after a fixed `N` chaining attempts. That feels a little more indirect to me, but I think that would also work fine here.

(Either way, I can document this in comments.)

Reply:
I thought about this a bit more, and I think we could achieve the same thing here by moving the `cert_is_self_signed` check into `build_chain_inner`, after the trust store check -- that _should_ be behaviorally equivalent, without the need to pre-iterate over the intermediate set. WDYT @alex?